build/ci_utils/src/io.rs

```diff
@@ -74,15 +74,33 @@ pub async fn download_and_extract(
     client::download_and_extract(&default(), url, output_dir).await
 }
 
-// pub async fn stream_to_file<E: Into<Box<dyn std::error::Error + Send + Sync>>>(
-//     stream: impl Stream<Item = std::result::Result<Bytes, E>> + Unpin,
-//     output_path: impl AsRef<Path>,
-// ) -> Result {
-//     let mut reader = tokio_util::io::StreamReader::new(stream.map_err(std::io::Error::other));
-//     let mut output = crate::fs::tokio::create(output_path).await?;
-//     tokio::io::copy(&mut reader, &mut output).await?;
-//     Ok(())
-// }
+/// Retry a given action until it succeeds or the maximum number of attempts is reached.
+pub async fn retry<Fn, Fut, Ret>(mut action: Fn) -> Result<Ret>
+where
+    Fn: FnMut() -> Fut,
+    Fut: Future<Output = Result<Ret>>, {
+    let growth_factor = 1.5;
+    let mut attempts = 5;
+    let mut delay = std::time::Duration::from_millis(500);
+    let max_delay = std::time::Duration::from_secs(10);
+
+    loop {
+        match action().await {
+            Ok(result) => return Ok(result),
+            Err(err) => {
+                let warning = format!("Failed to execute action: {err:?}");
```

Comment:
How about writing what operation failed to execute? Something like in https://github.com/enso-org/enso/blob/06c8d2977d3d998d094fbe91ab501096d4299b1d/app/gui/src/retry.rs#L79 or smarter

Reply:
> How about writing what operation failed to execute? Something like in 

I expect this to happen already by two possible ways:
* we pretty-print the error here, which should include not only the direct error but the whole stack explaining what was being attempted (thanks to `anyhow`);
* additionally caller might define a `tracing` scope (e.g. by annotating the function), so the logs from `retry` will be automatically assigned to the parent operation.